File: firmware/ci/buildAll.ts
import * as fs from "fs";
import glob from "glob";
import path from "path";
import { executeCommand, stringifyArray, uniqueArrayItems } from "./helpers";

process.chdir("..");

const puts = console.log;

const abortOnError = process.argv.includes("--abortOnError");
const failIfError = process.argv.includes("--failIfError");
const buildUpdatedOnly = process.argv.includes("--updatedOnly");

class BuildStepError extends Error {}

function getAllProjectVariationPaths() {
  return glob
    .sync("src/projects/**/rules.mk")
    .map((fpath) => path.dirname(fpath))
    .filter(
      (fpath) =>
        fs.existsSync(path.join(fpath, "config.h")) &&
        fs.existsSync(path.join(path.dirname(fpath), "project.json"))
    )
    .map((fpath) => path.relative("src/projects", fpath))
    .filter((vp) => !vp.startsWith("dev/"));
}

function getUpdatedProjectVariationPaths() {
  const [gitOutput] = executeCommand(
    `git diff origin/variants...HEAD --name-only`
  );
  const dirs = uniqueArrayItems(
    gitOutput.split(/\r?\n/).map((fpath) => path.dirname(fpath))
  );
  return dirs
    .filter((fpath) => fpath.startsWith("firmware/src/projects"))
    .map((fpath) => path.relative("firmware", fpath))
    .filter(
      (fpath) =>
        fs.existsSync(path.join(fpath, "rules.mk")) &&
        fs.existsSync(path.join(fpath, "config.h")) &&
        fs.existsSync(path.join(path.dirname(fpath), "project.json"))
    )
    .map((fpath) => path.relative("src/projects", fpath))
    .filter((vp) => !vp.startsWith("dev/"));
}

function makeFirmwareBuild(projectPath: string, variationName: string) {
  const command = `make ${projectPath}:${variationName}:build IS_RESOURCE_ORIGIN_ONLINE=1`;
  const [_stdout, stderr, status] = executeCommand(command);
  if (status !== 0) {
    throw new BuildStepError(`>${command}\n${stderr}`);
  }
}

function buildFirmware(projectPath: string, variationName: string): boolean {
  const targetName = `${projectPath}--${variationName}`;
  puts(`build ${targetName} ...`);
  // executeCommand(`make ${projectPath}:${variationName}:clean`);
  try {
    makeFirmwareBuild(projectPath, variationName);
    puts(`build ${targetName} ... OK`);
    return true;
  } catch (error) {
    if (error instanceof BuildStepError) {
      puts(error.message);
      if (abortOnError) {
        console.warn(`abort: failed to build ${targetName}`);
        process.exit(1);
      }
      puts(`build ${targetName} ... NG`);
      puts();
      return false;
    } else {
      console.error(error);
      process.exit(1);
    }
  }
}

function buildProjectVariation(projectVariationPath: string) {
  const projectPath = path.dirname(projectVariationPath);
  const variationName = path.basename(projectVariationPath);
  return buildFirmware(projectPath, variationName);
}

function buildProjects() {
  const projectVariationPaths = buildUpdatedOnly
    ? getUpdatedProjectVariationPaths()
    : getAllProjectVariationPaths();
  puts(`target project variations: ${stringifyArray(projectVariationPaths)}`);

  // executeCommand("make clean");
  const results = projectVariationPaths.map(buildProjectVariation);

  const numSuccess = results.filter((result) => !!result).length;
  const numTotal = results.length;
  if (numTotal > 0) {
    puts(`buildStats ${numSuccess}/${numTotal}`);
  }
  if (failIfError && numSuccess !== numTotal) {
    puts(`errors in some projects`);
    process.exit(1);
  }
  puts("done");
}

buildProjects();

File: firmware/workers/build_all.rb
require 'open3'
require 'pathname'

AbortOnError = ARGV.include?('--abortOnError')

def get_all_project_variation_paths
  Dir.glob('./src/projects/**/rules.mk')
     .map { |path| File.dirname(path) }
     .map { |path| path.sub('./src/projects/', '') }
     .select { |path| !(path.start_with?('dev') || path.start_with?('study')) }
     .sort
end

def make_firmware_build(project_path, variation_name)
  command = "make #{project_path}:#{variation_name}:build"
  _stdout, stderr, status = Open3.capture3(command)
  return { result: :ok } if status == 0

  { result: :ng, error_log: ">#{command}\n#{stderr}" }
end

def build_firmware(project_path, variation_name)
  puts "build #{project_path}--#{variation_name} ..."

  # `make #{project_path}:#{variation_name}:clean`
  res = make_firmware_build(project_path, variation_name)
  if res[:result] == :ok
    print "\e[A\e[K"
    puts "build #{project_path}--#{variation_name} ... OK"
    true
  else
    puts(res[:error_log])
    if AbortOnError
      warn("abort: failed to build #{project_path}--#{variation_name}")
      exit(1)
    end
    puts "build #{project_path}--#{variation_name} ... NG"
    puts
    false
  end
end

def build_project_variation(project_variation_path)
  project_path, variation_name = File.split(project_variation_path)
  build_firmware(project_path, variation_name)
end

def build_projects
  project_variation_paths = get_all_project_variation_paths
  # `make clean`
  puts "target project variations: #{project_variation_paths}"
  results = project_variation_paths.map { |it| build_project_variation(it) }
  num_success = results.count(true)
  num_total = results.length
  puts "build_stats #{num_success}/#{num_total}" if num_total > 0
  puts 'done'
end

build_projects if __FILE__ == $0
